### PR TITLE
chore(gatsby): Remove LOCKED_IN feature flags

### PR DESCRIPTION
**Pull request by LekoArts**
## Description

Removes `LOCKED_IN` feature flags.

- remove PRESERVE_WEBPACK_CACHE
- remove DEV_WEBPACK_CACHE
- remove LAZY_IMAGES
- remove LMDB_STORE
- remove PARALLEL_QUERY_RUNNING
- remove GRAPHQL_TYPEGEN
- remove QUERY_ON_DEMAND

### Documentation

No update needed since this is all internal. Only thing we could document is the new env var to enable QoD in CI (in the migration guide or something)

## Related Issues

[ch55428]

### CODE REVIEW

File: packages/gatsby/src/datastore/datastore.ts [hunk 1]
```diff
@@ -2,47 +2,15 @@ import { IDataStore } from "./types"
...
 export function getDataStore(): IDataStore {
   if (!dataStore) {
-    if (isLmdb) {
-      const { setupLmdbStore } = require(`./lmdb/lmdb-datastore`)
-      dataStore = setupLmdbStore()
-    } else {
-      const { setupInMemoryStore } = require(`./in-memory/in-memory-datastore`)
-      dataStore = setupInMemoryStore()
-    }
+    const { setupLmdbStore } = require(`./lmdb/lmdb-datastore`)
```
Comment (LekoArts) [inline packages/gatsby/src/datastore/datastore.ts:8]: Didn't delete the `in-memoy` directory yet. I guess we don't want to keep it?

File: packages/gatsby/src/datastore/datastore.ts [hunk 2]
```diff
@@ -2,47 +2,15 @@ import { IDataStore } from "./types"
...
 let dataStore: IDataStore
-let isLmdb = isLmdbStoreFlagSet()
 
 export function getDataStore(): IDataStore {
   if (!dataStore) {
-    if (isLmdb) {
-      const { setupLmdbStore } = require(`./lmdb/lmdb-datastore`)
-      dataStore = setupLmdbStore()
-    } else {
-      const { setupInMemoryStore } = require(`./in-memory/in-memory-datastore`)
```
Comment (LekoArts) [inline packages/gatsby/src/datastore/datastore.ts:13]: Didn't delete the `in-memoy` directory yet. I guess we don't want to keep it?

File: packages/gatsby/src/redux/__tests__/index.js [hunk 3]
```diff
@@ -221,212 +217,6 @@ describe(`redux db`, () => {
     })
   })
 
-  describe(`Sharding`, () => {
```
Comment (LekoArts) [inline packages/gatsby/src/redux/__tests__/index.js:224]: With LMDB enabled this never ran, so removed it

File: packages/gatsby/src/services/initialize.ts [hunk 4]
```diff
@@ -214,17 +207,15 @@ export async function initialize({
...
-    }
+  // Throughout the codebase GATSBY_QUERY_ON_DEMAND is used to conditionally enable the QoD behavior, depending on if "gatsby develop" is running or not. In CI QoD is disabled by default, too.
+  // You can use GATSBY_ENABLE_QOD_IN_CI to force enable it in CI.
+  process.env.GATSBY_QUERY_ON_DEMAND = `true`
+  if (process.env.gatsby_executing_command !== `develop`) {
+    // we don't want to ever have this flag enabled for anything than develop
+    // in case someone have this env var globally set
+    delete process.env.GATSBY_QUERY_ON_DEMAND
+  } else if (isCI() && !process.env.GATSBY_ENABLE_QOD_IN_CI) {
+    delete process.env.GATSBY_QUERY_ON_DEMAND
```
Comment (tyhopp) [inline packages/gatsby/src/services/initialize.ts:218]: Worth keeping the log? Something like (formatting may need adjustment):

```suggestion
    delete process.env.GATSBY_QUERY_ON_DEMAND
    reporter.verbose(
      `Query on Demand feature is not available in CI environment by default, continuing with eager query running. You can enable it by setting the GATSBY_ENABLE_QOD_IN_CI env var to a truthy value.`
    )
```

File: e2e-tests/development-runtime/package.json [hunk 5]
```diff
@@ -32,7 +32,7 @@
   "license": "MIT",
   "scripts": {
     "build": "gatsby build",
-    "develop": "cross-env CYPRESS_SUPPORT=y ENABLE_GATSBY_REFRESH_ENDPOINT=true GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND=y gatsby develop",
+    "develop": "cross-env CYPRESS_SUPPORT=y ENABLE_GATSBY_REFRESH_ENDPOINT=y GATSBY_ENABLE_QOD_IN_CI=y gatsby develop",
```
Comment (tyhopp) [inline e2e-tests/development-runtime/package.json:35]: I was confused what `GATSBY_ENABLE_QOD_IN_CI` did for some time while reviewing this.

We should probably avoid adding new acronyms to the vernacular where possible (more searchable, accessible to newcomers, etc.), even if the alternatives are more verbose.

Maybe something like `GATSBY_ENABLE_QUERY_ON_DEMAND_IN_CI` or `GATSBY_QUERY_ON_DEMAND_CI`?

File: packages/gatsby/src/datastore/datastore.ts (hunk 2 above)
Comment (LekoArts) [reply packages/gatsby/src/datastore/datastore.ts:13]: We still use `runFastFiltersAndSort` in LMDB for the `runQuery` without indexes https://github.com/gatsbyjs/gatsby/blob/remove-locked-in/packages/gatsby/src/datastore/lmdb/lmdb-datastore.ts#L209-L218

File: packages/gatsby/src/redux/__tests__/index.js (hunk 3 above)
Comment (pieh) [reply packages/gatsby/src/redux/__tests__/index.js:224]: we still do sharding for `pages` and not just `nodes` I think? nodes in redux are in fact empty, but pages are still in redux right now ( pages sharding introduced as part of https://github.com/gatsbyjs/gatsby/pull/28590 )

Comment (LekoArts) [reply packages/gatsby/src/redux/__tests__/index.js:224]: Good catch! So should I remove that early return then or rewrite the tests?

File: packages/gatsby/src/datastore/datastore.ts (hunk 2 above)
Comment (pieh) [reply packages/gatsby/src/datastore/datastore.ts:13]: We need to keep it. Not only lmdb indexes is not enabled by default, but also lmdb indexes just can't handle certain filters as well as filter+sort combinations - https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby/src/datastore/lmdb/query README might have a bit more information about details, but TL/DR is that even LMDB indexes might need to fallback to using fast filters

File: packages/gatsby/src/redux/__tests__/index.js (hunk 3 above)
Comment (pieh) [reply packages/gatsby/src/redux/__tests__/index.js:224]: probably remove early return and remove the setup / assertions for node sharding, but keep the setup/assertions for pages sharding ( I'm not sure if that would require more rewrites or not )

**Review by pieh — APPROVED** [review]
🎉

**Pull request merged**